include/odp/api/spec/pool.h

```diff
@@ -173,6 +173,107 @@ int odp_pool_stats(odp_pool_t pool, odp_pool_stats_t *stats);
  */
 int odp_pool_stats_reset(odp_pool_t pool);
 
+/**
+ * Query capabilities of an external memory pool type
+ *
+ * Outputs pool capabilities on success. Returns failure if the pool type is not supported as
+ * an external memory pool.
+ *
+ * @param      type    Pool type
+ * @param[out] capa    Pointer to capability structure for output
+ *
+ * @retval 0 on success
+ * @retval <0 on failure
+ */
+int odp_pool_ext_capability(odp_pool_type_t type, odp_pool_ext_capability_t *capa);
+
+/**
+ * Initialize pool params
+ *
+ * Initialize an odp_pool_ext_param_t to its default values for all fields
+ * based on the selected pool type.
+ *
+ * @param type     Pool type
+ * @param param    odp_pool_ext_param_t to be initialized
+ */
+void odp_pool_ext_param_init(odp_pool_type_t type, odp_pool_ext_param_t *param);
+
+/**
+ * Create an external memory pool
+ *
+ * This routine is used to create a pool. The use of pool name is optional.
+ * Unique names are not required. However, odp_pool_lookup() returns only a
+ * single matching pool. Use odp_pool_ext_param_init() to initialize parameters
+ * into their default values.
+ *
+ * @param name     Name of the pool or NULL. Maximum string length is ODP_POOL_NAME_LEN.
+ * @param param    Pool parameters
+ *
+ * @return Pool handle on success
+ * @retval ODP_POOL_INVALID on failure
+ */
+odp_pool_t odp_pool_ext_create(const char *name, const odp_pool_ext_param_t *param);
+
+/**
+ * Populate external memory pool with buffer memory
+ *
+ * Populate can be called multiple times to add memory buffers into the pool. Application must
+ * populate the pool with the exact number of buffers specified in pool parameters. The pool is
+ * ready to be used for allocations only after all populate calls have returned successfully.
+ * Application marks the last populate call with ODP_POOL_POPULATE_DONE flag.
+ *
+ * Depending on pool usage (and ODP implementation), the memory may need to be accessible by
+ * HW accelerators. Application may use e.g. odp_shm_reserve() with ODP_SHM_HW_ACCESS flag to
+ * ensure HW access. The memory area used for one pool (from the lowest addressed buffer to
+ * the end of the highest addressed buffer) must not overlap with the memory area used for
+ * any other pool.
```

Comment:
This still doesn't imply that sys pages must not be shared between pools. 

Could you add a capability "sys_page_sharing" described below (or add spec text which implies this restriction.)

"sys_page_sharing" capability when **not** set, puts a restriction that the memory area used for two different pools must not have common shared sys pages.

Memory area of a pool is defined as the memory between the beginning of lowest addressed buffer and the end of the highest addressed buffer populated into the pool.

sys page is defined as any memory area whose address is aligned to and having size of odp_sys_page_size().

Reply:
odp_sys_page_size() does not return e.g. "System MMU"  page size, but just regular page size of the OS (e.g. 4 or 64kB on Linux). A problem with "page sharing" flag is to define which page size it refers to: normal page, some huge page size, SMMU page or something else.

We could extend the above spec text with: "...must not overlap with memory area or memory pages used for any other pool." The same problem of what page size is meant would remain. We could also think that application should know memory mapping rules for this pool memory in this case, when it wants to do the memory management itself.

Memory area alignment requirement would allow implementation to specify the correct page size (from SMMU).

Reply:
>> Memory area alignment requirement would allow implementation to specify the correct page size (from SMMU).

Memory alignment would apply to the lowest addressed buffer being populated. Right? This in combination with **buf_size_aligned** flag (which is the case in Marvell implementation) would lead to lot of memory wastage.

Eg: Lets say we want a buffer size of 37KB. But if our system page size is 64KB, then we will have to do either of the following 2 things.
1) Find a page whose address is aligned to 37KB and 64KB to use as the first buffer. For this user will have to allocate a memory chunk of at least 37*64 KB to find such an address.
2) Give buffer size as 64KB but use only 37KB out of it.

Just another thought. May be add two additional parameters mem_area_start and mem_area_size which defines the range of buffer addresses that will be populated into the pool. Then add the 'Memory area alignment' capability which will apply to both mem_area_start and mem_area_size instead of the lowest addressed buffer.

Reply:
Actually, I did not mean alignment of the first buffer, but alignment of the memory area used for all the buffers: "minimum alignment of the start of the memory area used for an external memory pool".

Spec text could be formulated like this:
"The memory area used for one pool, staring from (or before) the lowest addressed buffer and 
extending to the end (or after) of the highest addressed buffer, must not overlap with the memory 
area or memory pages used for any other pool."

In your example, min_mem_align would be 64kB, and application would use that e.g. in odp_shm_reserve() call or reserve that much extra for start alignment.

Reply:
Okay.

If we define memory area as above and say that min_mem_align should apply to memory area, then I think it should be fine.

Reply:
Fixed in v14 with new memory area definition text, and added min_mem_align capa.